### ISSUE TITLE
mutual information loss

### ISSUE THREAD

**Issue by completementgaga** [kornia/kornia#3374]
## 🚀 Feature
A loss for mutual information between two images

## Motivation
The [mutual information](https://en.wikipedia.org/wiki/Mutual_information) is a generalization of Spearman correlation that can be valuable to define a loss function for image alignment.

## Pitch

 MILossFromReference would be a torch.nn.module

````
# reference_image shape : (H,W)
loss_from_ref = MILossFromReference(reference_image)
# input image must have shape (H,W)
loss_from_ref(input) 
````
would output a tensor which is a decreasing function of the mutual information between reference_image and input_image (so that minimizing this loss would help alignment)

We could have a mask optional argument for MILossFromReference, so as to ignore what happens in certain pixels.

## Alternatives

We could also implement mi_loss as a function of reference_image and input, but the module approach maybe allows to define dynamically a more efficient forward method in the no mask case, upon Module initialization. Isn't it?

______________________________________________________________________

**Comment by shijianjian** [comment]
In general, we can have both. A functional one for on-the-fly computations and a module one for getting rid of the initialization.

Masks shall be input for both functional and modules I think.

I am not sure if we should keep the reference image as an initialization argument. I prefer the following:
```
# reference_image shape : (H,W)
loss_from_ref = MILossFromReference()
# input image must have shape (H,W)
loss_from_ref(input, reference_image, mask) 
```

**Comment by completementgaga** [comment]
@shijianjian I suspect there will be some costly preprocessing of both images, so it would be better to do it only once for the reference image (upon instantiation), not at every training step.

**Comment by zer-art** [comment]
Hi @completementgaga and @shijianjian, I would like to work on this issue!
I have read the discussion regarding the API design. To satisfy both requirements, I propose implementing the core logic as a functional component (kornia.losses.mutual_information_loss) that takes both images as input. Then, I can implement the nn.Module wrapper.
Regarding the efficiency vs. flexibility debate: Could we implement the Module such that if a reference image is passed at __init__, it pre-calculates the stats? But if None is passed, it expects the reference in the forward() pass?
Let me know if you are happy for me to proceed with a PR.

**Comment by completementgaga** [comment]
@zer-art I suggest we wait for some orientation of the @kornia members. I am interested in contributing to the feature too. Maybe we can do it together.

**Comment by completementgaga** [comment]
I want to mention also a possible source of confusion: there is already a Kullback-Leibler divergence implemented in kornia.losses. Yet, in my understanding the Mutual information  would be the Kullback Leibler divergence between the density for (X,Y) and the product of  the densities for X and for Y, where X and Y are the images interpreted as random variables defined on Omega = support of the images = [0,H)x[0,W).

**Comment by completementgaga** [comment]
Actually the 2D aspect is almost irrelevant here, we could have a backend that works with the flattened data, so as to take benefit of the construct also for 3D images. However, to obtain relevant gradient here, I think we would need to bin the signal and make some Kernel Density Estimates.
The math seems to be seriously explained [here](https://www.researchgate.net/publication/220240108_Parzen-Window_Based_Normalized_Mutual_Information_for_Medical_Image_Registration).

**Comment by zer-art** [comment]
Hi @completementgaga, that sounds great! I would be happy to collaborate on this with you.

I agree we should wait for the maintainers' orientation on the API structure before finalizing the class design.

In the meantime, I can start prototyping the differentiable Kernel Density Estimation backend. Since we know we'll need a differentiable way to bin the signal (likely using Gaussian or triangular kernels on the flattened data), I can work on that core logic while we wait.

I'll review the resource you linked and draft a basic implementation.

**Comment by edgarriba** [comment]
Can you sketch here a skeleton of what you guys are proposing as API ?

**Comment by completementgaga** [comment]
@zer-art I just invited you as a collaborator to a kornia fork https://github.com/completementgaga/kornia.
I pushed my first implementation attempt that uses torch-kde. I think, we have to remove this dependency. Isn't it @edgarriba ?

**Comment by completementgaga** [comment]
> Can you sketch here a skeleton of what you guys are proposing as API ?

@edgarriba This loss would be added in kornia.losses, with core logic in kornia.losses.mutual_information. We would have the functional version mutual_information_loss and a torch.nn.Module version in the public API of kornia.losses.

Regarding the signatures, I imagine something as below.

````
def mutual_information_loss(reference_image:torch.Tensor,input:torch.Tensor)->torch.Tensor:
    module = MutualInformationFromRef(reference_image, **yet_undefined_supplementary_kwargs)
    return -module.forward(input)
````

where 
````
class MutualInformationFromRef(torch.nn.Module):
    def __init__(reference_image, **yet_undefined_supplementary_kwargs):

    def forward(input):
````

An alternative would be 
````
def mutual_information_loss(reference_image:torch.Tensor,input:torch.Tensor)->torch.Tensor:
    module = MutualInformation(**yet_undefined_supplementary_kwargs)
    return -module.forward(reference_image,input)
````

where 
````
class MutualInformation(torch.nn.Module):
    def __init__(**yet_undefined_supplementary_kwargs):

    def forward(reference_image,input):

````

The choice between the two alternatives should be led by a performance vs simplicity trade off, depending on the amount of preprocessing needed for reference_image.

Regarding the shapes: input, reference_image and mask should all have the same shape and would be all flattened before further processing.

**Comment by edgarriba** [comment]
Sounds good! Yes, please, send a PR whenever is ready to use

**Comment by completementgaga** [comment]
@kornia @edgarriba @shijianjian, as a matter of principles, can you give us Kornia's policy for dependency limitations? For instance, is it allowed to use torch-kde in Kornia ?

**Comment by zer-art** [comment]
@completementgaga I've pulled the changes and looked at the prototype.

Regarding the `torch-kde` dependency—I agree that we'll likely need to remove it to stick to Kornia's policy.

To proactively address this, I will start working on replacing the `torch-kde` dependency with a **native PyTorch implementation**. As discussed, I will implement the **differentiable Parzen Window estimation** using the 2nd-order polynomial kernel (from the Xu et al. paper) to ensure we get the correct gradient flow without needing external libraries.

I will push the changes directly to the `feat/mutual-information-loss` branch once I have the kernel logic working.

**Comment by completementgaga** [comment]
@zer-art 
Cool. In case you need It feel free to create subbranches. We can pursue the technical discussion on the fork's github, if you liké.

**Comment by edgarriba** [comment]
ideally it should be native pytorch to keep the library light from dependencies

**Comment by completementgaga** [comment]
I merged @zer-art's PR to my fork and the issue here was closed. Is It the expected behaviour? I find it quite strange.

**Comment by edgarriba** [comment]
why not sending directly in main kornia ?

**Comment by completementgaga** [comment]
> why not sending directly in main kornia ?

My idea was to polish things with zer-art on the fork and to make a PR to main kornia once we consider the feature branch would reach acceptable standards.

I think we are not there yet.

**Comment by zer-art** [comment]
I agree with @completementgaga. Since we identified some adjustments needed for gradient flow and normalization, it’s safer to stabilize it here on the fork first.

I’m currently writing the batch consistency tests he mentioned to verify that the vectorized implementation strictly matches the iterative approach. Once that is confirmed and the gradients are solid, we'll be ready to PR to main.

**Comment by completementgaga** [comment]
By the way, @edgarriba @shijianjian 
- regarding tests, is there a simple way to load example images in kornia, similar to skimage.data.camera() ?
- is there some customary way of sharing demo notebooks for the implemented features?

**Comment by edgarriba** [comment]
- for notebooks we have the tutorials repo https://github.com/kornia/tutorials as they are pretty heavy usually to be hosted in the main library repo
- as for image testing, ideally in the main library we do unit tests using synthetic image generation that meets the operator expectations; or generic samples to compare with other known libraries.

**Comment by zer-art** [comment]
Just ran the updated tests locally—batch consistency is passing perfectly now! ✅

Great work on the fix @completementgaga. The broadcasting logic is solid.

Regarding the Rectangular Kernels and nn.Module API: I think those are great ideas for optimization. To keep this PR focused on the stability fix, maybe we should merge this version to main first? I can definitely help implement those additional kernels in a follow-up PR!

I'll leave it to you to decide when you're ready to open the PR to Kornia.
